echopype/preprocess/api.py

```diff
@@ -329,3 +284,427 @@ def remove_noise(ds_Sv, ping_num, range_sample_num, noise_max=None, SNR_threshol
+    comp_er_check: bool = True,
+) -> np.ndarray:
+    """
+    Bins and means ``arr`` based on ``times`` and ``echo_range``,
+    and their corresponding bins. If ``arr`` is ``Sv`` then this
+    will compute the MVBS.
+
+    Parameters
+    ----------
+    arr: dask.array.Array or np.ndarray
+        The 2D array whose values should be binned
+    bins_time: np.ndarray
+        1D array (used by np.digitize) representing the binning required for ``times``
+    bins_er: np.ndarray
+        1D array (used by np.digitize) representing the binning required for ``echo_range``
+    times: np.ndarray
+        1D array corresponding to the time values that should be binned
+    echo_range: np.ndarray
+        2D array of echo range values
+    comp_er_check: bool
+        If True, a more comprehensive check will be completed to determine if ``echo_range``
+        grouping along ``ping_time`` is needed, otherwise a less comprehensive check will be done
+
+    Returns
+    -------
+    final_reduced: np.ndarray
+        The final binned and mean ``arr``, if ``arr`` is ``Sv`` then this is the MVBS
+
+    Notes
+    -----
+    This function assumes that ``arr`` has rows corresponding to
+    ``ping_time`` and columns corresponding to ``echo_range``.
+
+    This function should not be run if the number of ``echo_range`` values
+    vary amongst ``ping_times``.
+    """
+
+    # get the number of echo range and time bins
+    n_bin_er = len(bins_er)
+    n_bin_time = len(bins_time)
+
+    # obtain the bin indices for echo_range and times
+    digitized_echo_range, bin_time_ind = get_bin_indices(echo_range, bins_er, times, bins_time)
+
+    # determine if grouping of echo_range values with the same step size is necessary
+    if comp_er_check:
+        grouping_needed = is_grouping_needed_comprehensive(echo_range)
+    else:
+        grouping_needed = is_grouping_needed_less_comprehensive(echo_range)
+
+    if grouping_needed:
+        # groups, bins, and means arr with respect to echo_range
+        er_means = group_bin_mean_echo_range(arr, digitized_echo_range, n_bin_er)
+    else:
+        # bin and mean arr with respect to echo_range
+        er_means = bin_and_mean_echo_range(arr, digitized_echo_range[0, :], n_bin_er)
+
+    # if er_means is a dask array we compute it so the graph does not get too large
+    if isinstance(er_means, dask.array.Array):
+        er_means = er_means.compute()
+
+    # create final reduced array i.e. MVBS
+    final = np.empty((n_bin_time, n_bin_er - 1))
+    for bin_time in range(1, n_bin_time + 1):
+
+        # obtain er_mean indices corresponding to the time bin
+        indices = np.argwhere(bin_time_ind == bin_time).flatten()
+
+        if len(indices) == 0:
+            # fill values with NaN, if there are no values in the bin
+            final[bin_time - 1, :] = np.nan
+        else:
+            # bin and mean the er_mean time bin
+            final[bin_time - 1, :] = np.nanmean(er_means[:, indices], axis=1)
```

Comment:
I have a similar question as in under the function `bin_and_mean_echo_range` above. Wouldn't you be able to wrap the core ops into a function and delay it, so that the binned averages are distributed instead of computed in a loop?

Reply:
Or, `bin_and_mean_2d` can be delayed so that the MVBS in each channel is calculated in parallel?